File: src-macos/Sources/CRTCore/TimeFormatter.swift
import Foundation

/// Decimal helpers plus all time formatting, ported from `src/crt/decorators.py`.
///
/// Every time value in CRT is a Foundation `Decimal` so values such as 29.97
/// behave exactly. Rounding uses `NSDecimalRound` with `.bankers` (round half
/// to even) because that is what Python's `round(Decimal, n)` does by default
/// (`decimal.ROUND_HALF_EVEN`) and what the Windows port spells out as
/// `MidpointRounding.ToEven` — all three apps must agree on midpoints.
public enum TimeFormatter {

    // MARK: - Decimal helpers

    /// Rounds a decimal to `scale` fractional digits (round half to even,
    /// matching Python's default decimal context).
    public static func rounded(_ value: Decimal, scale: Int) -> Decimal {
        var input = value
        var result = Decimal()
        NSDecimalRound(&result, &input, scale, .bankers)
        return result
    }

    /// Rounds a decimal to the nearest integer and returns it as `Int`.
    public static func roundedToInt(_ value: Decimal) -> Int {
        let r = rounded(value, scale: 0)
        return NSDecimalNumber(decimal: r).intValue
    }

    /// Truncates a decimal toward zero and returns it as `Int`
    /// (the equivalent of Python's `int(Decimal)` cast).
    public static func truncatedToInt(_ value: Decimal) -> Int {
        var input = value
        var result = Decimal()
        NSDecimalRound(&result, &input, 0, value < 0 ? .up : .down)
        return NSDecimalNumber(decimal: result).intValue
    }

    /// Locale-independent plain string for a decimal (always '.' separator).
    public static func string(_ value: Decimal) -> String {
        return NSDecimalNumber(decimal: value).stringValue
    }

    // MARK: - Components (port of format_components)

    /// Numeric time components. Negative input clamps to zero.
    /// The milliseconds part is returned as a ready-made 3-digit string built
    /// by splitting the decimal string representation, exactly like the
    /// Python implementation does for precision-3 values.
    static func numericComponents(_ time: Decimal) -> (hours: Int, minutes: Int, seconds: Int, milliseconds: String) {
        let clamped = time < 0 ? Decimal(0) : time
        let text = string(clamped)

        let secondsPart: String
        var fractionPart: String
        if let dot = text.firstIndex(of: ".") {
            secondsPart = String(text[text.startIndex..<dot])
            fractionPart = String(text[text.index(after: dot)...])
        } else {
            secondsPart = text
            fractionPart = ""
        }

        // The fractional digits are positional: "05" means 50 ms. Pad on the
        // right to three digits — and cap at three, so a value carrying more
        // precision than the display format still renders exactly three
        // millisecond digits (spec §5).
        if fractionPart.count > 3 {
            fractionPart = String(fractionPart.prefix(3))
        } else {
            while fractionPart.count < 3 {
                fractionPart += "0"
            }
        }

        let totalSeconds = Int(secondsPart) ?? 0
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        return (hours, minutes, seconds, fractionPart)
    }

    /// String components in the Python `format_components` style:
    /// hours/minutes/seconds zero-padded to 2 digits, milliseconds 3 digits.
    public static func components(_ time: Decimal) -> (hours: String, minutes: String, seconds: String, milliseconds: String) {
        let (h, m, s, ms) = numericComponents(time)
        return (
            String(format: "%02d", h),
            String(format: "%02d", m),
            String(format: "%02d", s),
            ms
        )
    }

    // MARK: - ISO display format (port of format_iso)

    /// Drops leading zero *units*, but every unit that is shown is two digits:
    /// `SS.mmm` under a minute (zero state `"00.000"`), `MM:SS.mmm` under an
    /// hour (60s → `"01:00.000"`), else `HH:MM:SS.mmm` (3600s → `"01:00:00.000"`).
    /// Negative values clamp to zero. Milliseconds are always 3 digits.
    ///
    /// The leading unit stays zero-padded because the Python app interpolates
    /// its already-padded component strings, and mod notes produced here are
    /// pasted into speedrun.com by moderators — the apps must agree byte for
    /// byte. Verified against `src/crt/decorators.py`'s `format_iso`.
    public static func iso(_ time: Decimal) -> String {
        let (h, m, s, ms) = numericComponents(time)
        if h > 0 {
            return String(format: "%02d:%02d:%02d.", h, m, s) + ms
        }
        if m > 0 {
            return String(format: "%02d:%02d.", m, s) + ms
        }
        return String(format: "%02d.", s) + ms
    }

    /// Converts a frame count/position at the given framerate into an
    /// ISO-style timestamp. `"00.000"` when the framerate is zero.
    public static func frameTime(frames: Int, framerate: Decimal, precision: Int) -> String {
        if framerate == 0 {
            return iso(Decimal(0))
        }
        let seconds = rounded(Decimal(frames) / framerate, scale: precision)
        return iso(seconds)
    }

    /// Formats an absolute frame position as a YouTube chapter timestamp
    /// (`M:SS` or `H:MM:SS`) — chapters do not support milliseconds, the
    /// value floors to whole seconds.
    public static func youtubeTimestamp(frame: Int, framerate: Decimal) -> String {
        let totalSeconds: Int
        if framerate == 0 {
            totalSeconds = 0
        } else {
            totalSeconds = truncatedToInt(Decimal(frame) / framerate)
        }
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        }
        return String(format: "%d:%02d", minutes, seconds)
    }

    /// Speedrun.com style: `"HHh MMm SSs mmmms"` (port of src_format).
    public static func srcFormat(_ time: Decimal) -> String {
        let (h, m, s, ms) = components(time)
        return "\(h)h \(m)m \(s)s \(ms)ms"
    }
}

File: src-windows/CRT.Core/Formatting/TimeFormatter.cs
using System.Globalization;
using CRT.Core.Models;

namespace CRT.Core.Formatting;

/// <summary>
/// Time formatting ported from <c>src/crt/decorators.py</c>
/// (<c>format_components</c> / <c>format_iso</c> / <c>format_frame_time</c>).
/// </summary>
public static class TimeFormatter
{
    /// <summary>
    /// Splits a decimal seconds value into ("HH", "MM", "SS", "mmm") strings by
    /// decimal string splitting, exactly like the Python <c>format_components</c>.
    /// Negative values clamp to 0. Milliseconds are always 3 digits.
    /// </summary>
    public static (string Hours, string Minutes, string Seconds, string Milliseconds) FormatComponents(decimal time)
    {
        decimal clamped = time < 0m ? 0m : time;
        string timeStr = clamped.ToString(CultureInfo.InvariantCulture);

        long totalSeconds;
        string fraction;
        int dot = timeStr.IndexOf('.');
        if (dot >= 0)
        {
            totalSeconds = long.Parse(timeStr[..dot], CultureInfo.InvariantCulture);
            fraction = timeStr[(dot + 1)..];
        }
        else
        {
            totalSeconds = long.Parse(timeStr, CultureInfo.InvariantCulture);
            fraction = "";
        }

        // The Python code receives values quantized to exactly 3 decimal places
        // and does str(int(frac)).rjust(3, "0"). Normalizing the fractional digit
        // string to exactly 3 digits is equivalent for those values and also
        // correct for shorter-scale decimals (e.g. "65.1" → "100").
        string ms = fraction.Length >= 3 ? fraction[..3] : fraction.PadRight(3, '0');

        long minutes = Math.DivRem(totalSeconds, 60, out long seconds);
        long hours = Math.DivRem(minutes, 60, out minutes);

        return (
            hours.ToString("00", CultureInfo.InvariantCulture),
            minutes.ToString("00", CultureInfo.InvariantCulture),
            seconds.ToString("00", CultureInfo.InvariantCulture),
            ms);
    }

    /// <summary>
    /// ISO-style display format: drops leading zero <em>units</em>, but every unit
    /// that is shown is two digits — "SS.mmm" under a minute ("00.000" zero-state),
    /// "MM:SS.mmm" under an hour (60s → "01:00.000"), else "HH:MM:SS.mmm"
    /// (3600s → "01:00:00.000"). Negative clamps to 0.
    /// </summary>
    /// <remarks>
    /// The leading unit stays zero-padded because the Python app interpolates the
    /// already-padded component strings, and mod notes produced here are pasted
    /// into speedrun.com by moderators — the two apps must agree byte for byte.
    /// Verified against <c>src/crt/decorators.py</c>'s <c>format_iso</c>.
    /// </remarks>
    public static string FormatIso(decimal time)
    {
        var (hours, minutes, seconds, ms) = FormatComponents(time);
        long h = long.Parse(hours, CultureInfo.InvariantCulture);
        long m = long.Parse(minutes, CultureInfo.InvariantCulture);
        if (h > 0)
        {
            return $"{hours}:{minutes}:{seconds}.{ms}";
        }
        if (m > 0)
        {
            return $"{minutes}:{seconds}.{ms}";
        }
        return $"{seconds}.{ms}";
    }

    /// <summary>
    /// Converts a frame count/position at the given framerate into an ISO-style
    /// timestamp; "00.000" when framerate is 0.
    /// </summary>
    public static string FormatFrameTime(int frames, decimal framerate, int precision)
    {
        if (framerate == 0m)
        {
            return FormatIso(0m);
        }
        return FormatIso(Math.Round(frames / framerate, precision, MidpointRounding.ToEven));
    }

    /// <summary>
    /// YouTube chapter timestamp (M:SS or H:MM:SS — no milliseconds, floored to seconds).
    /// </summary>
    public static string FormatYouTubeTimestamp(int frame, decimal framerate)
    {
        long totalSeconds = framerate == 0m ? 0 : (long)(frame / framerate);
        long hours = Math.DivRem(totalSeconds, 3600, out long remainder);
        long minutes = Math.DivRem(remainder, 60, out long seconds);
        if (hours > 0)
        {
            return $"{hours}:{minutes:00}:{seconds:00}";
        }
        return $"{minutes}:{seconds:00}";
    }

    /// <summary>ISO display for the session's primary time ("without loads" / segment total).</summary>
    public static string PrimaryIso(TimeSession session) => FormatIso(session.PrimarySeconds);

    /// <summary>ISO display for the session's secondary time ("with loads" / full run).</summary>
    public static string SecondaryIso(TimeSession session) => FormatIso(session.SecondarySeconds);
}
